### ISSUE TITLE
Feedback on build documentation

### ISSUE THREAD

**Issue by rengolin** [plaidml/plaidml#1972]
Trying to build from the docs got me some figuring out, so I propose we update the docs to include the other stuff. But before I submit a docs change, I'd like to know what of those "figuring out" things are "commonly known", assumed or just me not getting it right.

So, here's a list of things I had to do in addition to the base docs:
1. After installing `conda`, I have created a script that enters/leaves the virtualenv, as I don't want to enter it on login (its default behaviour is to add it to `.bashrc`, which is silly). Basically, I _moved_ what `conda` added to `.bashrc` into a new script, and added two new aliases: `alias enter-plaidml='source plaidml-env-script'` and `alias leave-plaidml='conda deactivate'`.
2. To allow OpenMP testing, PlaidML needs `llvm-lit`, `FileCheck` and `not` on the PATH, but the `llvm-project-build` directory didn't have it. I have built LLVM+Clang+MLIR on my box and copied the files to a directory in the PATH (same as the env script). I don't know how that directory is pulled from, but it would be nice to have them there and maybe even automatically add it to the PATH.
3. I created an Ubuntu 22.04 container, and when running `./configure` I was told there's no support for OpenCV on that distro (even though I can install its packages, so I'm confused). The interim solution is to run `./configure --no_openvino`.
4. The docs don't mention submodules, but the two directories under `vendor` (`mlir-hlo` and `openvino`) don't get initialised by default. I suggest adding a step to `git submodule update --init --recursive` before calling `./configure`.
5. I had no `ccache` on my container, but the build system still chose to use it, then I got a `ccache: not found` when building. I just installed `ccache` (and `clang` and `lld` for good measure), but it would be good to only use it when it's on the PATH. Installing `clang` and adding `export CC=clang CXX=clang++` didn't make `configure` choose it for the build.

After the CMake is done, I get a nice message on how to run the unit tests, but the build has many error messages like:

```
../../plaidml/core/core.h:93:5: error: 'fn' cannot be used as a function
../../plaidml/core/core.h: In instantiation of 'void plaidml::ffi::call_void(F, Args ...) [with F = int; Args = {void (*)(plaidml_error*, plaidml_expr*, long unsigned int, plaidml_dim_expr**), plaidml_expr*, long unsigned int, plaidml_dim_expr**}]':
```

```
../../plaidml/op/lib/ops.cc: In function 'plaidml::edsl::Value plaidml::op::lib::variance(const plaidml::edsl::Value&)':
../../plaidml/op/lib/ops.cc:3629:33: error: no matching function for call to 'plaidml::edsl::Value::Value(<brace-enclosed initializer list>)'
 3629 |   return Value{SumSqDiff / denom};
      |                                 ^
```

```
../../plaidml/op/lib/ops.cc:3338:17: error: no matching function for call to 'plaidml::edsl::TensorDim::TensorDim()'
 3338 |       TensorDim G, C, K;
      |                 ^
```

```
/usr/include/c++/11/type_traits:955:52: error: static assertion failed: template argument must be a complete class or an unbounded array
  955 |       static_assert(std::__is_complete_or_unbounded(__type_identity<_Tp>{}),
      |                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~
```

IIUC, all errors have plenty of candidates, which means the compiler's ability to guess the types is going wrong. This is using `ccache c++` which defaults to `gcc`, and the version on my system is 11.2.0. I found no way to change it to clang.

I think the docs should mention how to create the environment that is known to work (ex. Ubuntu 20.04), so that people with newer OSs don't try to chase weird build errors. Creating a docker that has that environment would be a nice touch, too.

**Comment by chelini** [comment]
Hi @rengolin 

2. The directory is pulled here (https://github.com/plaidml/plaidml/blob/dc24b5804906191368126cf5f8eb259bb327d8f8/cmake/third_party/llvm-project.cmake#L37). The way I do it is to build LLVM and then make PlaidML point to it using `-local_llvm` in the configure file.

3. `./configure --no-openvino` is the right way. I have a PR in flight to remove it (https://github.com/plaidml/plaidml/pull/1958).

4 and 5. I agree 

Internally we use this:

```
export PLAIDML_WORKSPACE_DIR=[your choice]
mkdir -f ${PLAIDML_WORKSPACE_DIR}
cd ${PLAIDML_WORKSPACE_DIR}
 
wget https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh
bash Miniconda3-latest-Linux-x86_64.sh
eval "$(${PLAIDML_WORKSPACE_DIR}/miniconda3/bin/conda shell.bash hook)"
conda activate
 
git clone https://github.com/plaidml/plaidml.git -b plaidml-v1
git submodule update --recursive –init
OR
git clone https://github.com/plaidml/plaidml.git --recursive -b plaidml-v1
 
cd plaidml
conda env create -f environment.yml -p .cenv/
conda activate .cenv/
 
# preparing PlaidML build
./configure --no_openvino
#./configure --no_openvino --type=Debug
 
# buidling PlaidML
cd build-x86_64/Release
ninja && PYTHONPATH=$PWD python plaidml/plaidml_setup.py
```

@alheinecke knows more.

**Comment by alheinecke** [comment]
Sent a PM to @rengolin in parallel with the instructions you included here.

Fixed #1958 and merged it.
